### ISSUE TITLE
NoSuchElementException when TrustManagerFactory emits null TrustManager

### ISSUE THREAD

**Issue by kaeawc** [babylonhealth/certificate-transparency-android#35]
We're seeing this crash stacktrace when we attempted to rollout `0.2.0` to our app, looks to be happening for about 6% of users across all Android OS versions. We stopped the rollout for now and removed the certificatetransparency dependency until this can be addressed. I haven't been able to reproduce this error, we're only seeing it in the wild.

```
Fatal Exception: java.util.NoSuchElementException
       at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:366)
       at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:416)
       at java.util.ServiceLoader$1.next(ServiceLoader.java:494)
       at kotlin.collections.CollectionsKt___CollectionsKt.firstOrNull(CollectionsKt___CollectionsKt.java:224)
       at kotlin.collections.CollectionsKt.firstOrNull(CollectionsKt.java)
       at com.babylon.certificatetransparency.chaincleaner.CertificateChainCleaner$Companion.get(CertificateChainCleaner.java:27)
       at com.babylon.certificatetransparency.internal.verifier.CertificateTransparencyBase$cleaner$2.invoke(CertificateTransparencyBase.java:63)
       at com.babylon.certificatetransparency.internal.verifier.CertificateTransparencyBase$cleaner$2.invoke(CertificateTransparencyBase.java:42)
       at kotlin.SynchronizedLazyImpl.getValue(SynchronizedLazyImpl.java:74)
       at com.babylon.certificatetransparency.internal.verifier.CertificateTransparencyBase.hasValidSignedCertificateTimestamp(CertificateTransparencyBase.java:96)
       at com.babylon.certificatetransparency.internal.verifier.CertificateTransparencyBase.verifyCertificateTransparency(CertificateTransparencyBase.java:76)
       at com.babylon.certificatetransparency.internal.verifier.CertificateTransparencyInterceptor.intercept(CertificateTransparencyInterceptor.java:51)
       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:112)
       at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:37)
```

**Comment by kaeawc** [comment]
I think I've found the actual problem: CertificateChainCleaner uses a lazy ServiceLoader called `factoryLoader`, and when `fun get(trustManager: X509TrustManager)` is called the first thing that happens is `factoryLoader.firstOrNull()?.get(trustManager) ?: BasicCertificateChainCleaner(trustManager)`. From the point of view of the IDE this looks fine and if there are no elements in the factoryLoader we'll just use the `BasicCertificateChainCleaner`... but in actuality `ServiceLoader` is an `Iterable` and not a `List`, so the Kotlin `firstOrNull` behaves differently:

```
/**
 * Returns the first element, or `null` if the collection is empty.
 */
public fun <T> Iterable<T>.firstOrNull(): T? {
    when (this) {
        is List -> {
            if (isEmpty())
                return null
            else
                return this[0]
        }
        else -> {
            val iterator = iterator()
            if (!iterator.hasNext())
                return null
            return iterator.next()
        }
    }
}
```

The ServiceLoader is apparently returning true for `hasNext` because it's implementation assumes it has a next if `knownProviders` does without checking if `lookupIterator` does, but then throws the `NoSuchElement` exception when `iterator.hasNext()` is called.

@mattmook I think I would be able to solve this if CertificateChainCleaner interface was separated from the internal default implementation and we can pass in our own implementation. In that case then we could make use of OkHttp's `Platform` implementation for CertificateChainCleaner. What are your thoughts?

**Comment by kaeawc** [comment]
I can put together a PR that makes the CertificateChainCleaner an interface whose implentation uses the current one as default and otherwise customizable, does that sound okay to you?

**Comment by mattmook** [comment]
Apologies for the delay in replying, only just starting to catch up on this project.

The issue is caused by the service loader, actually before the trust manager factory is even called?

Service Loader adds an entry to META-INF/services, so there is a possibility that is either not there or invalid, causing the above. Were you actually able to replicate this on a device or emulator?

To be honest I've been meaning to switch the implementation away from the use of the ServiceLoader anyway and move to use plain reflection as it is significantly faster on Android.

I should have a PR in the next day or so that applies this change along with allowing an override so people can provide their own implementation, however, note that OkHttps implementation is supposed to be internal so you wouldn't be expected to call it directly (not that anything stops you, this library originally used it!)

**Comment by mattmook** [comment]
Hopefully fixed by #44
